File: letsPollApi/src/main/java/com/serverless/letspoll/handlers/PollGetter.java
package com.serverless.letspoll.handlers;

import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;
import com.serverless.letspoll.models.ApiGatewayResponse;
import com.serverless.letspoll.models.Poll;
import com.serverless.letspoll.models.generated.tables.records.PollRecord;
import org.jooq.DSLContext;
import com.serverless.letspoll.commons.DatabaseAccessUtils;


import java.util.Arrays;
import java.util.Collections;
import java.util.Map;


public class PollGetter implements RequestHandler<Map<String,Object>, ApiGatewayResponse> {

    public ApiGatewayResponse handleRequest(Map<String,Object>input, Context context) {
        Map<String, String> pathParameters = (Map<String, String>) input.get("pathParameters");
        String pollId = pathParameters.get("pollId");
        System.out.println("pollId passed is "+ pollId);

        DSLContext dslContext = DatabaseAccessUtils.getDatabaseConnection();
        PollRecord pollRecord = dslContext.fetchOne(
            com.serverless.letspoll.models.generated.tables.Poll.POLL,
            com.serverless.letspoll.models.generated.tables.Poll.POLL.POLL_ID.eq(pollId));

        if (pollRecord!=null) {
            Poll poll = new Poll();
            poll.setPollId(pollRecord.getPollId());
            poll.setPollTitle(pollRecord.getPollTitle());
            poll.setPollQuestion(pollRecord.getPollQuestion());
            return ApiGatewayResponse.builder().setStatusCode(200)
                .setObjectBody(poll)
                .setHeaders(Collections.singletonMap("X-Powered-By", "AWS Lambda & serverless"))
                .build();
        }
        else {
            return ApiGatewayResponse.builder().setStatusCode(404)
                .setObjectBody("The requested Poll Doesnt exist")
                .setHeaders(Collections.singletonMap("X-Powered-By", "AWS Lambda & serverless"))
                .build();
        }
    }
}

File: letsPollApi/src/main/kotlin/com/serverless/letspoll/handlers/PollGetter.kt
package com.serverless.letspoll.handlers

import com.amazonaws.services.lambda.runtime.Context
import com.amazonaws.services.lambda.runtime.RequestHandler
import com.serverless.letspoll.commons.DatabaseAccessUtils
import com.serverless.letspoll.models.ApiGatewayResponse
import com.serverless.letspoll.models.Poll
import com.serverless.letspoll.models.generated.tables.records.PollRecord
import com.serverless.letspoll.models.responses.PollDetailsResponse
import com.serverless.letspoll.models.responses.PollResponseStatistics
import com.serverless.letspoll.models.responses.RespondentDetails
import java.util.*


class PollGetter : RequestHandler<Map<String, Any>, ApiGatewayResponse> {

    override fun handleRequest(input: Map<String, Any>, context: Context): ApiGatewayResponse {
        val pathParameters = (input["pathParameters"] as? Map<String, String>)
        val pollId = pathParameters?.get("pollId")

        println("pollId passed is $pollId")

        val dslContext = DatabaseAccessUtils.getDatabaseConnection()
        dslContext?.let {
            try {
                val pollRecord: PollRecord? = it
                        .fetchOne(com.serverless.letspoll.models.generated.tables.Poll.POLL,
                                com.serverless.letspoll.models.generated.tables.Poll.POLL.POLL_ID.eq(pollId))

                val respondentRecord = it
                        .fetchOne(com.serverless.letspoll.models.generated.tables.Respondent.RESPONDENT,
                                com.serverless.letspoll.models.generated.tables.Respondent.RESPONDENT.A_RESPONDENT_ID
                                        .eq(pollRecord!!.createdBy))

                return if (pollRecord != null) {
                    val poll = Poll(pollRecord.pollId, pollRecord.pollTitle, pollRecord.pollQuestion)

                    val respondentDetails = RespondentDetails(respondentDisplayName = respondentRecord.respondentDisplayName,
                            respondentEmail = respondentRecord.respondentEmailId)

                    val stateOne = PollResponseStatistics("NO", 10)

                    val stateTwo = PollResponseStatistics("YES", 90)

                    val pollDetailsResponse = PollDetailsResponse(poll, respondentDetails, listOf(stateOne, stateTwo))

                    ApiGatewayResponse.builder().setStatusCode(200).setObjectBody(pollDetailsResponse)
                            .setHeaders(Collections.singletonMap("X-Powered-By", "AWS Lambda & serverless"))
                            .build()
                } else {
                    ApiGatewayResponse.builder().setStatusCode(404)
                            .setObjectBody("The requested Poll with id doesnt exist")
                            .setHeaders(Collections.singletonMap("X-Powered-By", "AWS Lambda & serverless"))
                            .build()
                }
            } catch (e: Exception) {
                e.printStackTrace()
                return ApiGatewayResponse.builder().setStatusCode(409)
                        .setObjectBody("Could not fetch the poll details")
                        .setHeaders(Collections.singletonMap("X-Powered-By", "AWS Lambda & serverless"))
                        .build()

            }

        }
        return ApiGatewayResponse.builder().setStatusCode(409)
                .setObjectBody("Could not fetch the poll details")
                .setHeaders(Collections.singletonMap("X-Powered-By", "AWS Lambda & serverless"))
                .build()
    }
}
